### PR TITLE
ARROW-2275: [C++] Guard against bad use of Buffer.mutable_data()

### PR DESCRIPTION
**Pull request by pitrou**
Also disambiguate the Tensor API on this front.

### CODE REVIEW

**Review by xhochy — APPROVED** [review]
+1, LGTM

File: cpp/src/arrow/buffer.cc [hunk 1]
```diff
@@ -70,6 +70,14 @@ Status Buffer::FromString(const std::string& data, std::shared_ptr<Buffer>* out)
   return FromString(data, default_memory_pool(), out);
 }
 
+#ifndef NDEBUG
+// DCHECK macros aren't allowed in public include files
+uint8_t* Buffer::mutable_data() {
+  DCHECK(is_mutable());
+  return mutable_data_;
+}
+#endif
```
Comment (wesm) [inline cpp/src/arrow/buffer.cc]: This is a good idea, thank you

File: cpp/src/arrow/buffer.h [hunk 2]
```diff
@@ -54,7 +54,11 @@ class ARROW_EXPORT Buffer {
   ///
   /// \note The passed memory must be kept alive through some other means
   Buffer(const uint8_t* data, int64_t size)
-      : is_mutable_(false), data_(data), size_(size), capacity_(size) {}
+      : is_mutable_(false),
+        data_(data),
+        mutable_data_(nullptr),
```
Comment (wesm) [inline cpp/src/arrow/buffer.h]: We need to use the NULLPTR macro here

**Review by wesm — APPROVED** [review]
I found one small issue, I will fix and then merge

File: cpp/src/arrow/buffer.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/buffer.h]: Can you expand on the NULLPTR macro? Does it do something more then `nullptr`?

Comment (wesm) [reply cpp/src/arrow/buffer.h]: `nullptr` is incompatible C++/CLI, which I believe is a way for C# code to link to C++ libraries. see https://msdn.microsoft.com/en-us/library/4ex65770.aspx

**Pull request closed without merging**